sense.go

```diff
@@ -0,0 +1,220 @@
+func (r *Relations) GoString() string {
+	return r.String()
+}
+
+// Set parse a string into a Relation and append it to the Relations slice if it is missing
+// this is used by the flag package
+func (r *Relations) Set(value string) error {
+	relation, err := NewRelation(value)
+	if err != nil {
+		return err
+	}
+	r.addIfMissing(relation)
+
+	return nil
+}
+
+func (r *Relations) addIfMissing(value *Relation) {
+	for _, relation := range *r {
+		if relation.Equal(value) {
+			return
+		}
+	}
+
+	*r = append(*r, value)
+}
+
+// AddToIndex add the relations to the specified category on the index
+func (r *Relations) AddToIndex(category relationCategory, h *index) error {
+	var (
+		nameTag,
+		versionTag,
+		flagsTag int
+		num      = len(*r)
+		names    = make([]string, num)
+		versions = make([]string, num)
+		flags    = make([]uint32, num)
+	)
+
+	if num == 0 {
+		return nil
+	}
+
+	switch category {
+	case ProvidesCategory:
+		nameTag = tagProvides
+		versionTag = tagProvideVersion
+		flagsTag = tagProvideFlags
+	case RequiresCategory:
+		nameTag = tagRequires
+		versionTag = tagRequireVersion
+		flagsTag = tagRequireFlags
+	case ObsoletesCategory:
+		nameTag = tagObsoletes
+		versionTag = tagObsoleteVersion
+		flagsTag = tagObsoleteFlags
+	case SuggestsCategory:
+		nameTag = tagSuggests
+		versionTag = tagSuggestVersion
+		flagsTag = tagSuggestFlags
+	case RecommendsCategory:
+		nameTag = tagRecommends
+		versionTag = tagRecommendVersion
+		flagsTag = tagRecommendFlags
+	case ConflictsCategory:
+		nameTag = tagConflicts
+		versionTag = tagConflictVersion
+		flagsTag = tagConflictFlags
+	default:
+		return fmt.Errorf("unknown category %s", category)
+	}
+
+	for idx := range *r {
+		relation := (*r)[idx]
+		names[idx] = relation.Name
+		versions[idx] = relation.Version
+		flags[idx] = uint32(relation.Sense)
+	}
+
+	h.Add(nameTag, entry(names))
+	h.Add(versionTag, entry(versions))
+	h.Add(flagsTag, entry(flags))
+
+	return nil
+}
+
+// NewRelation parse a string into a Relation
+func NewRelation(related string) (*Relation, error) {
+	var (
+		err   error
+		sense rpmSense
+	)
+	parts := relationMatch.FindStringSubmatch(related)
+	if sense, err = parseSense(parts[2]); err != nil {
+		return nil, err
+	}
+
+	return &Relation{
+		Name:    parts[1],
+		Version: parts[3],
+		Sense:   sense,
+	}, nil
+}
+
+var senseStrings = map[rpmSense]string{
+	SenseAny:                  "",
+	SenseLess:                 "<",
+	SenseGreater:              ">",
+	SenseEqual:                "=",
+	SenseLess | SenseEqual:    "<=",
+	SenseGreater | SenseEqual: ">=",
+}
+
+// String return the string representation of the rpmSense
+func (r rpmSense) String() string {
+	var (
+		ok  bool
+		ret string
+	)
+
+	if ret, ok = senseStrings[r]; !ok {
+		return "UNKNOWN"
+	}
+
+	return ret
+}
+
+// GoString return the string representation of the rpmSense
+func (r rpmSense) GoString() string {
+	return r.String()
+}
+
+func parseSense(sense string) (rpmSense, error) {
```

Comment:
Let's pivot senseStrings to be a mapping from `string` to `rpmSense`, and make the searching here trivial.